# Tarea 1

## Instrucciones

Utilizar la secuencia almacenada en la variable `vibrio_cholerae` y encontrar:
- Cadena complementaria
- Conteo de cada base
- Imprimir los conteos
- Calcular que porcentaje de cada base hay en la secuencia
- Escribir en markdown los resultados

In [1]:
vibrio_cholerae = "atcaatgatcaacgtaagcttctaagcatgatcaaggtgctcacacagtttatccacaacctgagtggatgacatcaagataggtcgttgtatctccttcctctcgtactctcatgaccacggaaagatgatcaagagaggatgatttcttggccatatcgcaatgaatacttgtgacttgtgcttccaattgacatcttcagcgccatattgcgctggccaaggtgacggagcgggattacgaaagcatgatcatggctgttgttctgtttatcttgttttgactgagacttgttaggatagacggtttttcatcactgactagccaaagccttactctgcctgacatcgaccgtaaattgataatgaatttacatgcttccgcgacgatttacctcttgatcatcgatccgattgaagatcttcaattgttaattctcttgcctcgactcatagccatgatgagctcttgatcatgtttccttaaccctctattttttacggaagaatgatcaagctgctgctcttgatcatcgtttc"

In [2]:
print(vibrio_cholerae)

atcaatgatcaacgtaagcttctaagcatgatcaaggtgctcacacagtttatccacaacctgagtggatgacatcaagataggtcgttgtatctccttcctctcgtactctcatgaccacggaaagatgatcaagagaggatgatttcttggccatatcgcaatgaatacttgtgacttgtgcttccaattgacatcttcagcgccatattgcgctggccaaggtgacggagcgggattacgaaagcatgatcatggctgttgttctgtttatcttgttttgactgagacttgttaggatagacggtttttcatcactgactagccaaagccttactctgcctgacatcgaccgtaaattgataatgaatttacatgcttccgcgacgatttacctcttgatcatcgatccgattgaagatcttcaattgttaattctcttgcctcgactcatagccatgatgagctcttgatcatgtttccttaaccctctattttttacggaagaatgatcaagctgctgctcttgatcatcgtttc


## Diccionario de complementos 

<img src="https://www.genome.gov/sites/default/files/media/images/tg/Base-pair_dyn.jpg">

### Cadena complementaria & Conteos de bases

El objetivo de las siguientes celdas es generar el codigo necesario para calcular la cadena complementaria y contar las bases adentro de un ciclo `for`.

#### Paso 1:
Vamos a crear un diccionario de Python que tenga como `Key` la letra que representa a la base y como `value` su complemento.

In [3]:
##  Tu codigo aqui :
complementos = {
    'a':'t',
    't':'a',
    'c':'g',
    'g':'c',
}

In [4]:
print(complementos)

{'a': 't', 't': 'a', 'c': 'g', 'g': 'c'}


In [5]:
base = "t"
complementos[base]

'a'

In [6]:
complementos["a"]

't'

In [7]:
#complementos["u"]

In [8]:
complementos.get("u", "kjhkj")

'kjhkj'

In [9]:
complementos.get("u")

#### Paso 2:
Vamos a utilizar la variable `base` del ciclo para buscar en el diccionario su complemento en el diccionario, una vez obtenemos el complemento hay que concatenarlo en la variable `cadena_complementaria`.
Utilizamos de nuevo la variable `base` y utilizando un `if` vamos a ir sumando en las variables `conteo_` 

In [10]:
cadena_complementaria = ""

cadena = "aatc"

# En estas variables vamos a ir almacenando los conteos
base_cont = {'a': 0, 't': 0, 'c': 0, 'g': 0}

#Obtener el complemento del diccionario y concatenarlo a la candena "cadena_complementaria"
cadena_complementaria = ''.join(complementos[base] for base in vibrio_cholerae)

## Contar que tipo de base es: en cada iteracion sumanos uno a la variable dependiendo del tipo de base que es
for base in vibrio_cholerae:
    base_cont[base] += 1



In [11]:
print(vibrio_cholerae)

atcaatgatcaacgtaagcttctaagcatgatcaaggtgctcacacagtttatccacaacctgagtggatgacatcaagataggtcgttgtatctccttcctctcgtactctcatgaccacggaaagatgatcaagagaggatgatttcttggccatatcgcaatgaatacttgtgacttgtgcttccaattgacatcttcagcgccatattgcgctggccaaggtgacggagcgggattacgaaagcatgatcatggctgttgttctgtttatcttgttttgactgagacttgttaggatagacggtttttcatcactgactagccaaagccttactctgcctgacatcgaccgtaaattgataatgaatttacatgcttccgcgacgatttacctcttgatcatcgatccgattgaagatcttcaattgttaattctcttgcctcgactcatagccatgatgagctcttgatcatgtttccttaaccctctattttttacggaagaatgatcaagctgctgctcttgatcatcgtttc


In [12]:
print(cadena_complementaria)

tagttactagttgcattcgaagattcgtactagttccacgagtgtgtcaaataggtgttggactcacctactgtagttctatccagcaacatagaggaaggagagcatgagagtactggtgcctttctactagttctctcctactaaagaaccggtatagcgttacttatgaacactgaacacgaaggttaactgtagaagtcgcggtataacgcgaccggttccactgcctcgccctaatgctttcgtactagtaccgacaacaagacaaatagaacaaaactgactctgaacaatcctatctgccaaaaagtagtgactgatcggtttcggaatgagacggactgtagctggcatttaactattacttaaatgtacgaaggcgctgctaaatggagaactagtagctaggctaacttctagaagttaacaattaagagaacggagctgagtatcggtactactcgagaactagtacaaaggaattgggagataaaaaatgccttcttactagttcgacgacgagaactagtagcaaag


#### Conteos

In [13]:
## Imprime aqui los conteos de las bases 
for key,val in base_cont.items():
    print(f'La cadena {key} estuvo {val} veces')

La cadena a estuvo 136 veces
La cadena t estuvo 174 veces
La cadena c estuvo 122 veces
La cadena g estuvo 108 veces


### Calcular que porcentaje de cada base hay en la secuencia

Crea una función en python llamada `get_base_percentage()`, que reciba como parametros
la longiutd total de la secuencia, el conteo de la base y el nombre de la base.

Ejemplo: La función debe retornar lo siguiente:
Llamada a la función 
```python
get_base_percentage(10, 100, "adenina")
```
Salida:
```
"adenina: 10%"
```


In [14]:

def get_base_percentage(longitud_total, conteo_base, nombre):
    percentage = (conteo_base / longitud_total ) * 100
    return f'{nombre} : {round(percentage,2)} %'.upper()

In [15]:
longitud_total = len(vibrio_cholerae)
print(longitud_total)

540


In [16]:
type_cadenas = ['a','t','c','g']
cadenas_name = {'a': 'Adenina', 't': 'timina', 'c': 'citosina', 'g': 'guanina'}
for cad in type_cadenas:
    print(get_base_percentage(longitud_total, base_cont[cad], cadenas_name[cad]))

ADENINA : 25.19 %
TIMINA : 32.22 %
CITOSINA : 22.59 %
GUANINA : 20.0 %


# Resultados

En esta sección se pide redactar un pequeño parrafo con los resultados obtenidos anteriormente en una nueva celda de Markdown. Tratar de darle formato utilizando titulos, listas, etc.

[Markdown Cheatsheet](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)

> Puntos extra en la tarea si escriben los conteos de las bases en una tabla de markdown

| Cadena   | %      |
| -------- | ------ |
| ADENINA  | 25.19  |
| TIMINA   | 32.22  |
| CITOSINA | 22.59  |
| GUANINA  | 20.0  |
  |
  |

## Resultados
Encontre ...

- ADENINA : 25.19 %'
- CITOSINA : 22.59 %
- TIMINA  : 32.22 %'
- GUANINA : 20.00 %

# Validando  los Resultados

Esta seccion solo hay que ejecutarla no es necesario cambiar nada, con esto podemos validar los resultados

In [17]:
import Bio
from Bio.Seq import Seq

In [18]:
dna = Seq(vibrio_cholerae)

In [19]:
dna.count("a")

136

In [20]:
dna.count("t")

174

In [21]:
dna.count("c")

122

In [22]:
dna.count("g")

108

In [23]:
str(dna.complement())

'tagttactagttgcattcgaagattcgtactagttccacgagtgtgtcaaataggtgttggactcacctactgtagttctatccagcaacatagaggaaggagagcatgagagtactggtgcctttctactagttctctcctactaaagaaccggtatagcgttacttatgaacactgaacacgaaggttaactgtagaagtcgcggtataacgcgaccggttccactgcctcgccctaatgctttcgtactagtaccgacaacaagacaaatagaacaaaactgactctgaacaatcctatctgccaaaaagtagtgactgatcggtttcggaatgagacggactgtagctggcatttaactattacttaaatgtacgaaggcgctgctaaatggagaactagtagctaggctaacttctagaagttaacaattaagagaacggagctgagtatcggtactactcgagaactagtacaaaggaattgggagataaaaaatgccttcttactagttcgacgacgagaactagtagcaaag'

### Puntos Extra: 
Investigar en la documentación de Biopython que para que sirve la función `Seq`, que tipo de objeto retorna y documentarlo en una celda en esta notebook utilizando markdown


Documentacion de Biopython: [Documentacion](https://biopython.org/)

### Validando cadena complementaria

In [24]:
try:
    assert cadena_complementaria == str(dna.complement())
    print("Resultado correcto!!!")
except AssertionError as msg:
    print("La cadena complementaria es incorrecta")

Resultado correcto!!!


## Validando conteos

In [25]:
try:
    assert base_cont['a'] == dna.count("a")
    print("Conteo a correcto")
except AssertionError as msg:
    print("conteo incorrecto")

Conteo a correcto


In [26]:
try:
    assert base_cont['t'] == dna.count("t")
    print("Conteo a correcto")
except AssertionError as msg:
    print("conteo incorrecto")

Conteo a correcto


In [27]:
try:
    assert base_cont['c'] == dna.count("c")
    print("Conteo a correcto")
except AssertionError as msg:
    print("conteo incorrecto")

Conteo a correcto


In [28]:
try:
    assert base_cont['g'] == dna.count("g")
    print("Conteo a correcto")
except AssertionError as msg:
    print("conteo incorrecto")

Conteo a correcto
